### DISCUSSION TITLE
[10월 1주차] 하샤드 수

### DISCUSSION THREAD

**Discussion by z1nun** [z1nun/ASTERISK-#27] [알고리즘 문제풀이]
<h6 class="guide-section-title" style="box-sizing: border-box; padding: 0px; margin: 0px 0px 1rem; color: white; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-variant-numeric: inherit; font-variant-east-asian: inherit; font-weight: 700; font-stretch: inherit; font-size: 14px; line-height: 1.5; font-family: Inter, NotoSansKR, &quot;Helvetica Neue&quot;, Helvetica, Arial, &quot;맑은 고딕&quot;, &quot;malgun gothic&quot;, 돋움, Dotum, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Noto Color Emoji&quot;; -webkit-font-smoothing: antialiased; letter-spacing: normal; orphans: 2; text-align: left; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(38, 55, 71); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">문제 설명</h6><div class="markdown solarized-dark" style="box-sizing: border-box; padding: 0px; margin: 0px; color: rgb(178, 192, 204); font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-variant-numeric: inherit; font-variant-east-asian: inherit; font-weight: 400; font-stretch: inherit; font-size: 16px; line-height: inherit; font-family: Inter, NotoSansKR, &quot;Helvetica Neue&quot;, Helvetica, Arial, &quot;맑은 고딕&quot;, &quot;malgun gothic&quot;, 돋움, Dotum, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Noto Color Emoji&quot;; background-color: rgb(38, 55, 71); word-break: normal; overflow-wrap: break-word; letter-spacing: normal; orphans: 2; text-align: left; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;"><p style="box-sizing: border-box; padding: 0px; margin: 0px; color: inherit; font-style: inherit; font-variant: inherit; font-weight: 400; font-stretch: inherit; font-size: inherit; line-height: 1.8; font-family: Inter, NotoSansKR, &quot;Helvetica Neue&quot;, Helvetica, Arial, &quot;맑은 고딕&quot;, &quot;malgun gothic&quot;, 돋움, Dotum, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Noto Color Emoji&quot;; letter-spacing: -0.1px; overflow-wrap: break-word;">양의 정수 x가 하샤드 수이려면 x의 자릿수의 합으로 x가 나누어져야 합니다. 예를 들어 18의 자릿수 합은 1+8=9이고, 18은 9로 나누어 떨어지므로 18은 하샤드 수입니다. 자연수 x를 입력받아 x가 하샤드 수인지 아닌지 검사하는 함수, solution을 완성해주세요.</p><h5 style="box-sizing: border-box; padding: 0px; margin: 1.5rem 0px 0px; color: white; font-style: inherit; font-variant: inherit; font-weight: 400; font-stretch: inherit; font-size: 15px; line-height: 1.5; font-family: Inter, NotoSansKR, &quot;Helvetica Neue&quot;, Helvetica, Arial, &quot;맑은 고딕&quot;, &quot;malgun gothic&quot;, 돋움, Dotum, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Noto Color Emoji&quot;; letter-spacing: -0.009em;">제한 조건</h5><ul style="box-sizing: border-box; padding: 0px 0px 0px 2.5rem; margin: 0.75rem 0px 0px; color: inherit; font: inherit;"><li style="box-sizing: border-box; padding: 0px; margin: 0px; color: inherit; font-style: inherit; font-variant: inherit; font-weight: inherit; font-stretch: inherit; font-size: 16px; line-height: 1.6; font-family: inherit; letter-spacing: -0.009em; list-style: inherit;"><code style="box-sizing: border-box; padding: 2px 6px; margin: 0px 2px 0px 0px; color: rgb(205, 215, 224); font-style: inherit; font-variant: inherit; font-weight: inherit; font-stretch: inherit; font-size: 14px; line-height: inherit; font-family: Hack, Menlo, Monaco, Consolas, &quot;Courier New&quot;, monospace, &quot;맑은 고딕&quot;, &quot;malgun gothic&quot;, 돋움, Dotum, sans-serif; overflow-wrap: break-word; border: 1px solid rgb(23, 35, 52); background-color: rgb(32, 43, 61); border-radius: 0.25rem;">x</code>는 1 이상, 10000 이하인 정수입니다.</li></ul><h5 style="box-sizing: border-box; padding: 0px; margin: 1.5rem 0px 0px; color: white; font-style: inherit; font-variant: inherit; font-weight: 400; font-stretch: inherit; font-size: 15px; line-height: 1.5; font-family: Inter, NotoSansKR, &quot;Helvetica Neue&quot;, Helvetica, Arial, &quot;맑은 고딕&quot;, &quot;malgun gothic&quot;, 돋움, Dotum, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Noto Color Emoji&quot;; letter-spacing: -0.009em;">입출력 예</h5>

arr | return
-- | --
10 | true
12 | true
11 | false
13 | false

<h5 style="box-sizing: border-box; padding: 0px; margin: 1.5rem 0px 0px; color: white; font-style: inherit; font-variant: inherit; font-weight: 400; font-stretch: inherit; font-size: 15px; line-height: 1.5; font-family: Inter, NotoSansKR, &quot;Helvetica Neue&quot;, Helvetica, Arial, &quot;맑은 고딕&quot;, &quot;malgun gothic&quot;, 돋움, Dotum, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Noto Color Emoji&quot;; letter-spacing: -0.009em;">입출력 예 설명</h5><p style="box-sizing: border-box; padding: 0px; margin: 0.75rem 0px 0px; color: inherit; font-style: inherit; font-variant: inherit; font-weight: 400; font-stretch: inherit; font-size: inherit; line-height: 1.8; font-family: Inter, NotoSansKR, &quot;Helvetica Neue&quot;, Helvetica, Arial, &quot;맑은 고딕&quot;, &quot;malgun gothic&quot;, 돋움, Dotum, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Noto Color Emoji&quot;; letter-spacing: -0.1px; overflow-wrap: break-word;"><strong style="box-sizing: border-box; padding: 0px; margin: 0px; color: inherit; font-style: inherit; font-variant: inherit; font-weight: bolder; font-stretch: inherit; font-size: inherit; line-height: inherit; font-family: inherit; -webkit-font-smoothing: antialiased;">입출력 예 #1</strong><br style="box-sizing: border-box; padding: 0px; margin: 0px; color: inherit; font: inherit;">10의 모든 자릿수의 합은 1입니다. 10은 1로 나누어 떨어지므로 10은 하샤드 수입니다.</p><p style="box-sizing: border-box; padding: 0px; margin: 0.75rem 0px 0px; color: inherit; font-style: inherit; font-variant: inherit; font-weight: 400; font-stretch: inherit; font-size: inherit; line-height: 1.8; font-family: Inter, NotoSansKR, &quot;Helvetica Neue&quot;, Helvetica, Arial, &quot;맑은 고딕&quot;, &quot;malgun gothic&quot;, 돋움, Dotum, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Noto Color Emoji&quot;; letter-spacing: -0.1px; overflow-wrap: break-word;"><strong style="box-sizing: border-box; padding: 0px; margin: 0px; color: inherit; font-style: inherit; font-variant: inherit; font-weight: bolder; font-stretch: inherit; font-size: inherit; line-height: inherit; font-family: inherit; -webkit-font-smoothing: antialiased;">입출력 예 #2</strong><br style="box-sizing: border-box; padding: 0px; margin: 0px; color: inherit; font: inherit;">12의 모든 자릿수의 합은 3입니다. 12는 3으로 나누어 떨어지므로 12는 하샤드 수입니다.</p><p style="box-sizing: border-box; padding: 0px; margin: 0.75rem 0px 0px; color: inherit; font-style: inherit; font-variant: inherit; font-weight: 400; font-stretch: inherit; font-size: inherit; line-height: 1.8; font-family: Inter, NotoSansKR, &quot;Helvetica Neue&quot;, Helvetica, Arial, &quot;맑은 고딕&quot;, &quot;malgun gothic&quot;, 돋움, Dotum, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Noto Color Emoji&quot;; letter-spacing: -0.1px; overflow-wrap: break-word;"><strong style="box-sizing: border-box; padding: 0px; margin: 0px; color: inherit; font-style: inherit; font-variant: inherit; font-weight: bolder; font-stretch: inherit; font-size: inherit; line-height: inherit; font-family: inherit; -webkit-font-smoothing: antialiased;">입출력 예 #3</strong><br style="box-sizing: border-box; padding: 0px; margin: 0px; color: inherit; font: inherit;">11의 모든 자릿수의 합은 2입니다. 11은 2로 나누어 떨어지지 않으므로 11는 하샤드 수가 아닙니다.</p><p style="box-sizing: border-box; padding: 0px; margin: 0.75rem 0px 0px; color: inherit; font-style: inherit; font-variant: inherit; font-weight: 400; font-stretch: inherit; font-size: inherit; line-height: 1.8; font-family: Inter, NotoSansKR, &quot;Helvetica Neue&quot;, Helvetica, Arial, &quot;맑은 고딕&quot;, &quot;malgun gothic&quot;, 돋움, Dotum, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Noto Color Emoji&quot;; letter-spacing: -0.1px; overflow-wrap: break-word;"><strong style="box-sizing: border-box; padding: 0px; margin: 0px; color: inherit; font-style: inherit; font-variant: inherit; font-weight: bolder; font-stretch: inherit; font-size: inherit; line-height: inherit; font-family: inherit; -webkit-font-smoothing: antialiased;">입출력 예 #4</strong><br style="box-sizing: border-box; padding: 0px; margin: 0px; color: inherit; font: inherit;">13의 모든 자릿수의 합은 4입니다. 13은 4로 나누어 떨어지지 않으므로 13은 하샤드 수가 아닙니다.</p></div>

https://school.programmers.co.kr/learn/courses/30/lessons/12947

**Comment by z1nun** [comment]
```js
function solution(x) {
    var answer = true;
    
    // 하샤드 수이려면 x의 자리수의 합으로 x가 나누어져야한다.
    // 18 -> 1+8=9 -> 18 % 9 === 0 -> True
    let num = 0;
    // x는 정수 이기 때문에 먼저 문자열로 바꿔준다
    const s = String(x)
    for (i of s) {
        num = num + Number(i) 
    }
    x % num === 0 ? answer : answer = false;
    
    
    
    return answer;
}

**Comment by sweetyr928** [comment]
```js
function solution(x) {
    return !(x % x.toString().split('').reduce((acc,cur) => acc+Number(cur),0))
}
```

**Comment by 2Zerozero** [comment]
하샤드 수
```
function solution(x) {
    let answer = true;
    let result = '';
    // 1+8 = 9 18은 9로 나누어 떨어지니 true
    // 2+9 = 11 29는 11로 나누어 떨이지지 않으니 false
    let stringX = String(x);
    let sumX = 0;
    // Stirng 타입으로 만들어서, 배열로 모든자리수의 합을 구할것이다.
    for(let i=0; i < stringX.length; i++) {
        result = result + stringX[i];
        sumX = sumX + Number(stringX[i]);
        if(Number(result) % Number(sumX) === 0) {
            answer = true;
        }else {
            answer = false;
        }
    }
    return answer;
}
```

**Comment by iltae** [comment]
```js
function solution(x) {
    // 숫자를 문자로 바꾸어 배열로 쪼개고 그 요소가 문자열인 숫자이기 때문에 숫자로 바꾸어 reduce함수로 더한다.
    // 입력 받은 수를 그 수로 나눌 때 나머지가 0 이면 true를 반환한다.
    return x % x.toString().split('').reduce((acc,cur) => acc + Number(cur), 0) === 0
}
```

**Comment by rhino-ty** [comment]
```js
function solution(x) {
  // 1. 합 구하기
  let sum = x
    .toString()
    .split("")
    .reduce((acc, cur) => acc + Number(cur), 0);
  // 2. 나누기
  return !(x % sum);
}
```

**Comment by fairytale779** [comment]
```js
function solution(x) {
    let har = String(x).split('').reduce((acc,cur)=>Number(acc)+Number(cur),0);
    return (x % har === 0) ? true : false 
}

**Comment by FordangIT** [comment]
```js
function solution(x) {
    let arr = String(x).split('') 
    // x를 문자로 바꾼후, split을 이용해 배열에 각각 넣는다.
    let number = 0; 
    for(let el of arr){
        number += Number(el); 
    }
    return x%number === 0 ? true : false
}
```
